src/snowflake/connector/config_parser/parser.py

```diff
@@ -0,0 +1,188 @@
+
+_T = TypeVar("_T")
+
+LOGGER = logging.getLogger(__name__)
+
+
+class ConfigOption:
+    def __init__(
+        self,
+        name: str,
+        _root_parser: ConfigParser,
+        _nest_path: list[str],
+        _type: Callable[[str], _T] | None = None,
+        choices: Iterable[_T] | None = None,
+        env_name: str | None | Literal[False] = None,
+    ) -> None:
+        """Create a config option that can read values from different locations.
+
+        Args:
+            name: The name of the ConfigOption
+            env_name: Environmental variable value should be read from, if not supplied, we'll construct this
+            type: A function that can turn str to the desired type, useful for reading value from environmental variable
+        """
+        self.name = name
+        self.type = _type
+        self.choices = choices
+        self._nest_path = _nest_path + [name]
+        self._root_parser: ConfigParser = _root_parser
+        self.env_name = env_name
+
+    def get(self):
+        """Retrieve a value of option."""
+        source = "environment variable"
+        loaded_env, value = self._get_env()
+        if not loaded_env:
+            source = "configuration file"
+            value = self._get_config()
+        if self.choices and value not in self.choices:
+            raise ConfigSourceError(
+                f"The value of {self.generate_name()} read from "
+                f"{source} is not part of {self.choices}"
+            )
+        return value
+
+    def generate_name(self) -> str:
+        return ".".join(self._nest_path[1:])
+
+    def generate_env_name(self) -> str:
+        pieces = map(lambda e: e.upper(), self._nest_path[1:])
+        return f"SF{'_' + '_'.join(pieces)}"
+
+    def _get_env(self) -> tuple[bool, str | _T | None]:
+        if self.env_name is False:
+            return False, None
+        if self.env_name is not None:
+            env_name = self.env_name
+        else:
+            # Generate environment name if it wasn't not explicitly supplied,
+            #  and isn't disabled
+            env_name = self.generate_env_name()
+        if env_name not in os.environ:
+            return False, None
+        env_var = os.environ.get(env_name, None)
+        loaded_var: _T | None | str = env_var
+        if env_var and self.type is not None:
+            loaded_var = self.type(env_var)
+        if isinstance(loaded_var, (Table, TOMLDocument)):
+            # If we got a TOML table we probably want it in dictionary form
+            return True, loaded_var.value
+        return True, loaded_var
+
+    def _get_config(self):
+        e = self._root_parser._conf
+        for k in self._nest_path[1:]:
+            e = e[k]
+        if isinstance(e, (Table, TOMLDocument)):
+            # If we got a TOML table we probably want it in dictionary form
+            return e.value
+        return e
+
+
+class ConfigParser:
+    def __init__(
+        self,
+        *,
+        name: str,
```

Comment:
I just want to call it out that `name` here is a required keyword argument, we are not making it a positional argument?

Reply:
Since this is a public API I just thought that it'd be a better idea to require all usages to specify every argument as a named argument. It makes things less confusing.

Reply:
I don't in general think that all arguments of public APIs need to be keyword-only arguments.  It can make for more verbose call sites.  That said, there are some good use cases for them (e.g. boolean arguments).  Sumit and I were just discussing this particular style recommendation.